### DISCUSSION TITLE
[API Request] Ability to set and get ratings on AlbumID3 and ArtistID3 types

### DISCUSSION THREAD

**Discussion by dweymouth** [opensubsonic/open-subsonic-api#17] [API requests]
### Type of change

API extension

### Proposal description

Some servers may already support this but officially in the last Subsonic API spec, only songs and music folders (Album and Artist, but not AlbumID3 and ArtistID3) could have ratings. We should extend setRating to support IDs for ArtistID3 and AlbumID3 entities, and return the rating in the JSON/XML response in the getAlbum and getArtist calls.

### Backward compatibility

- [X] No backward compatibility impact.

### API details

As described above. Add `userRating` (edit: and `averageRating`) as properties to the AlbumID3 and ArtistID3 types.

**Comment by Tolriq** [comment]
Beware of the meaning of ratings. There's ratings like the one you get from last.fm and there's userRatings.

Both should be supported and clearly disctincted.

**Reply by deluan** [reply]
AFAIK, Ratings in the Subsonic API only means userRatings, it does not fetch/submit ratings to Last.fm or any other external service. If we want to support that, we need to create an extension for that.

**Reply by dweymouth** [reply]
Tolriq is right though, that the property name should be `userRating`, not `rating` to keep it in line with the rest of the API. There is the concept of `averageRating` for album and artist (directories only, not ID3) and we probably want to add that as well to the ID3 types.

**Reply by ghenry22** [reply]
> AFAIK, Ratings in the Subsonic API only means userRatings, it does not fetch/submit ratings to Last.fm or any other external service. If we want to support that, we need to create an extension for that.

The original subsonic does not submit anything anywhere.  The ratings are purely user ratings as you say.

the ability to set and get ratings is really super limited.

I would propose to run through the api and look at all the places it would be useful.  For instance when fetching random or by genre or other criteria having the option get highly rated or apply a rating filter of above or below certain stars would be excellent.

**Comment by holocronweaver** [comment]
**What would be required to add album and artist rating APIs?**

Would love to see this API added. I rarely rate individual songs, but I almost always rate albums (and sometimes artists in the event they don't release albums). Historically most professional music ratings are based on albums. While the web has made standalone singles more common, for certain artists and genres, albums are still coherent pieces of work that can be rated as a whole. Album ratings are usually where I start when narrowing down what to listen to or delete. I'm surprised OpenSubsonic doesn't already support this.

**Comment by deluan** [comment]
I think we could introduce a new `userRating` attribute in all elements (folders, artists, albums and songs/child), and it could be a float with .5 steps, so it would support half-stars as well.

This could be a simple addition. What do you guys think?

**Reply by kgarner7** [reply]
1. Why not 1-10 (easier to represent, same result)? Or, if you'd like to support broader values, 1-100?
2. You would naturally need a new extension to specify support for this
3. Does a client send **both** 1-5 **and** the new scale when starting? Or only one variant? In the latter, then you'd need a new endpoint
4. How does the 10-point (or any other scale) convert back to the current 5-point scale? Presumably you would want the new scale to also be reflected on old clients. Floor down? Floor up? Round (100-point scale)?

**Reply by deluan** [reply]
1. Consistency. The existing [`averageRating`](https://subsonic.org/pages/inc/api/schema/subsonic-rest-api-1.16.1.xsd#:~:text=%3Cxs%3AsimpleType%20name%3D%22AverageRating%22%3E) is already float from 0 to 5: 
2. Yep, it is required if we want to be able set the rating.
3. Yes, similar to how we send `artist` and all the new `displayArtist`. The old one is meant for older unmaintained clients
4. Open for suggestions :)

**Reply by kgarner7** [reply]
1. I'm not too keen on floats in general (even if it is for consistency), but _only_ half steps feels worse (then you have to decide, is 1.5 1 or 2).
2. No comments there
3. I assume by that you mean a new **endpoint** to only set the new rating.
4. That depends ultimately on the implementation (and more broadly, how clients would display it). If it's functionally a larger range than 1-10, (so, 1-100, or infinite if a float), then round (with the exception that anything < 1, but greater than 0 is a 1). If it's functionally equivalent to a 1-10 scale, maybe ceiling?

**Reply by Tolriq** [reply]
IMO floats are bad if you want to express the 0...5 range with only half star support. As float should allow 1.1 2.3 and then showing as stars becomes a nightmare. (And json does not have a difference between float/double/...)
Some clients may display the whole 10 stars too. 
For consistency on the client side and logic 0..10 is better choice for ratings with half stars.

For the conversion to old for 0..10 it's commonly ceiling in most app or other providers.

**Reply by dweymouth** [reply]
What about adding a new OpenSubsonic field `userRatingAddHalfStar`? If true, it means that the rating is `userRating` + 0.5 (e.g. if `userRating` is 4, then the actual rating is 4.5 stars). It's backward compatible since we don't need to change `userRating` at all, and avoids the rounding problem of float.

**Reply by lachlan-00** [reply]
userRating should stay int 0-5 and averageRating stay float.

maybe adding something like userRatingFloat as an element or userPOPM with a larger scale would allow you to ignore the originals

**Reply by TGlide** [reply]
Any update on this? Would love to see half-ratings !! Thanks for all the hard work.

**Comment by RndOrc** [comment]
I would also love to have half star ratings as an option. The 5 star ratings scale (which is really only 4 stars if 0 stars = null) is just about doable for my purposes when I combine it with the favorite option, but having a bit more options available would help a great deal.

**Comment by Cr1tias** [comment]
what's the status on this? it's been debated for almost three years. I would also like for half star ratings to be supported to put subsonic servers in line with most music players

**Comment by SpaXter25** [comment]
Any updates on this? Especially half stars are still widely requested.

**Comment by bootybigboi** [comment]
Also bumping this, half star ratings are dearly desired. Adding a new boolean `userRatingAddHalfStar` seems like an easy, simple, backward compatible implementation.

**Comment by certuna** [comment]
**Apps/APIs supporting half stars:**
- Plex
- MusicBee
- Media Monkey
- Kodi
- Apple Music (not by default anymore, but enabled with `defaults write com.apple.music allow-half-stars -bool TRUE`)

**Apps/APIs only supporting 1-5 stars:**
- Subsonic
- VirtualDJ
- djay Pro
- Rekordbox

**Comment by starkgate** [comment]
Would also love to see half star ratings. I'm really missing that feature from MusicBee.

**Comment by Jolls** [comment]
New here, but am in the process of leaving mediamonkey for navidrome and I whole heartedly support half stars or smaller even. 0-100 would be my preference.  but would be happy with half stars. all my smart playlists are based on my ratings. As I only use 3+ stars in MM currently (which is 5 levels) I am considering mapping those to 1-5 (which is only 4 levels) to maintain my smart playlists. It'll be weird to listen to a 1 star song, but i'll know what it means.
